Unit tests

dosvath opened the pull request:
This resolves issue #1 
Added unit tests for grpcat.

dosvath committed:
Added unit tests

dosvath committed:
Added instructions for unit tests

adamf commented on README.md:
```diff
@@ -21,6 +21,32 @@ Assumptions
+
+Run the following way:
+
+`$ cat get_hello_request.txt | unit_tests.py`
```
We shouldn't require the user to create the file to run the unit tests. You should instead:

Create a hello.proto:
```
package grpcat;

message Hello {
  required string name = 1;
}
```

Compile it, and add the proto file and the resulting generated module to this branch. 

Then, in order to call grpcat as a module inside the unit tests, you would use a mock of fileinput.input to set standard in to the string that would be in get_hello_request.txt. See python 3's `unittest` and `mock` modules (and the `@patch` decorator) to see how to do this. This way the user doesn't have to supply the file to the unit test.

dosvath replied:
In the new commit I created the hello.proto and its compiled modules, implemented the mock fileinput.input, and created a sample get_hello_request.txt

adamf commented on unit_tests.py:
```diff
@@ -0,0 +1,167 @@
+
+
+
+    def test_successful_connection(self):
```
How will this test ever pass if there's no server?

dosvath replied:
The test checks whether grpcat raises any exceptions, if there are no exceptions raised by grpcat, there should be a response from the server. We could add the mock server for testing, or the test could be there for testing once there is a server?

adamf replied:
We should mock the server. It'll be better in the long run.

dosvath replied:
For the mock server would it be better to create a `mock_server.py` or start it on a thread in the `unit_tests.py`?

adamf commented on unit_tests.py:
```diff
@@ -0,0 +1,167 @@
+            self.assertEqual(expected_parse[i], str(retval))
+
+
+    def test_arg_count_errors(self):
```
This test fails for me:

```
======================================================================
FAIL: test_arg_count_errors (__main__.TestArgumentParsing)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "unit_tests.py", line 86, in test_arg_count_errors
    self.assertEqual(err.getvalue(),expected_errors[i])
AssertionError: u'' != 'usage: unit_tests.py [-h] host port rpc message_type\nunit_tests.py: error: the following arguments are required: message_type\n'
```

dosvath replied:
Did you run using python3? It runs on my side...

adamf replied:
Ah, you are right - sorry!

Commit:
Update README.md

dosvath committed:
Added an example hello.proto along with its compiled modules.
Added a sample get_hello_request.txt

dosvath committed:
Merge remote-tracking branch 'origin/unit_tests' into unit_tests

Commit:
Added a mock_server.py for the unit tests that test connectivity.
Minor fixes: added missing package to hello.proto, fixed test_failed_connection test in case of no failed exception.

dosvath committed:
Added mock_server to git, updated readme

Commit:
Updated README

dosvath committed:
Revised the implementation of the unit tests to not require the mock_server.py
The tests now run independently an mock the server connections for the successful and failed connection tests.

Updated test usage in README

adamf reviewed (commented):
Please run unit_tests.py to through a PEP-8 formatter; after that this will look good to commit. We'll have some post-commit issues to open, but I don't want to delay this landing.

adamf committed:
Update README.md
